public/app/features/alerting/unified/rule-list/PaginatedDataSourceLoader.tsx

```diff
@@ -0,0 +1,128 @@
+import { groupBy } from 'lodash';
+import { useEffect, useMemo, useRef } from 'react';
+
+import { Dropdown, Icon, IconButton, Menu, Stack, Text } from '@grafana/ui';
+import { DataSourceRuleGroupIdentifier, ExternalRulesSourceIdentifier, RuleGroup } from 'app/types/unified-alerting';
+
+import { hashRule } from '../utils/rule-id';
+
+import { AlertRuleLoader } from './AlertRuleLoader';
+import { DataSourceSection, DataSourceSectionProps } from './components/DataSourceSection';
+import { LazyPagination } from './components/LazyPagination';
+import { ListGroup } from './components/ListGroup';
+import { ListSection } from './components/ListSection';
+import { usePrometheusGroupsGenerator } from './hooks/prometheusGroupsGenerator';
+import { usePaginatedPrometheusGroups } from './hooks/usePaginatedPrometheusGroups';
+
+export const GROUP_PAGE_SIZE = 40;
+
+interface PaginatedDataSourceLoaderProps extends Required<Pick<DataSourceSectionProps, 'application'>> {
+  rulesSourceIdentifier: ExternalRulesSourceIdentifier;
+}
+export function PaginatedDataSourceLoader({ rulesSourceIdentifier, application }: PaginatedDataSourceLoaderProps) {
+  const { uid, name } = rulesSourceIdentifier;
+  const prometheusGroupsGenerator = usePrometheusGroupsGenerator();
+
+  const groupsGenerator = useRef(prometheusGroupsGenerator(rulesSourceIdentifier, GROUP_PAGE_SIZE));
+
+  useEffect(() => {
+    const currentGenerator = groupsGenerator.current;
+    return () => {
+      currentGenerator.return();
+    };
+  }, [groupsGenerator]);
+
+  const {
+    page: groupsPage,
+    nextPage,
+    previousPage,
+    canMoveForward,
+    canMoveBackward,
+    isLoading,
+  } = usePaginatedPrometheusGroups(groupsGenerator.current, GROUP_PAGE_SIZE);
+
+  const groupsByNamespace = useMemo(() => groupBy(groupsPage, 'file'), [groupsPage]);
+
+  return (
+    <DataSourceSection name={name} application={application} uid={uid} isLoading={isLoading}>
```

Comment:
same here, do you think we could create a reusable component for data source and grafana managed loaders?